exchanges/kline/kline.go

```diff
@@ -0,0 +1,128 @@
+package kline
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+)
+
+// CreateKline creates candles out of trade history data for a set time interval
+func CreateKline(trades []order.TradeHistory, interval time.Duration, p currency.Pair, a asset.Item, exchange string) (Item, error) {
+	if interval < time.Minute {
+		return Item{}, fmt.Errorf("invalid time interval: [%s]", interval)
+	}
+
+	err := validateData(trades)
+	if err != nil {
+		return Item{}, err
+	}
+
+	timeIntervalStart := trades[0].Timestamp.Truncate(interval)
+	timeIntervalEnd := trades[len(trades)-1].Timestamp
+
+	// Adds time interval buffer zones
+	var timeIntervalCache [][]order.TradeHistory
+	var OpenClose []HeartBeat
+
+	for t := timeIntervalStart; t.Before(timeIntervalEnd); t = t.Add(interval) {
+		timeBufferEnd := t.Add(interval)
+		insertionCount := 0
+
+		var zonedTradeHistory []order.TradeHistory
+		for i := 0; i < len(trades); i++ {
+			if (trades[i].Timestamp.After(t) ||
+				trades[i].Timestamp.Equal(t)) &&
+				(trades[i].Timestamp.Before(timeBufferEnd) ||
+					trades[i].Timestamp.Equal(timeBufferEnd)) {
+				zonedTradeHistory = append(zonedTradeHistory, trades[i])
+				insertionCount++
+				continue
+			}
+			trades = trades[i:]
+			break
+		}
+
+		// Insert dummy in time period when there is no price action
+		if insertionCount == 0 {
+			OpenClose = append(OpenClose, HeartBeat{Open: t, Close: timeBufferEnd})
+			timeIntervalCache = append(timeIntervalCache, []order.TradeHistory{})
+			continue
+		}
+		OpenClose = append(OpenClose, HeartBeat{Open: t, Close: timeBufferEnd})
+		timeIntervalCache = append(timeIntervalCache, zonedTradeHistory)
+	}
+
+	var candles = Item{
+		Exchange: exchange,
+		Pair:     p,
+		Asset:    a,
+		Interval: interval,
+	}
+
+	var closePriceOfLast float64
+	for x := range timeIntervalCache {
+		if len(timeIntervalCache[x]) == 0 {
+			candles.Candles = append(candles.Candles, Candle{
+				Time:  OpenClose[x].Open,
+				High:  closePriceOfLast,
+				Low:   closePriceOfLast,
+				Close: closePriceOfLast,
+				Open:  closePriceOfLast})
+			continue
+		}
+
+		var newCandle Candle
+		for y := range timeIntervalCache[x] {
+			if y == 0 {
```

Comment:
This can probably be done outside of this loop since you already check the len above. Then you only do this once